File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
// first solution (two loops)

    /*ListNode* reverseList(ListNode* head) {
        ListNode* wp = head;
        int temp;
        while(wp != NULL){
            ListNode* next = wp->next;
            while(next != NULL){
                temp = wp->val;
                wp->val = next->val;
                next->val = temp;
                next = next->next;
            }
            wp = wp->next;
        }
        return head;
    }*/

    // second solution
    int size(ListNode* head){
        ListNode* wp = head;
        int res = 0;
        while(wp!=NULL){
            res++;
            wp = wp->next;
        }
        return res;
    }
    ListNode* reverseList(ListNode* head) {

        // find size
        int listSize = size(head);

        // init vector with all zeros
        vector<int> v(size(head), 0);
        ListNode* wp = head;

        // write elements to the vector in reverse order
        int i = 0;
        while(wp!=NULL){
            v[listSize-1-i] = wp->val;
            i++;
            wp = wp->next;
        }

        // write elements from the vector to the list
        wp = head;
        i = 0;
        while(wp!=NULL){
            wp->val = v[i];
            i++;
            wp = wp->next;
        }
        return head;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    // first solution

    /*public ListNode reverseList(ListNode head) {
        int temp;
        ListNode wp = head;
        while(wp!=null){
            ListNode wpn = wp.next;
            while(wpn!=null){
                temp = wp.val;
                wp.val = wpn.val;
                wpn.val = temp;

                wpn=wpn.next;
            }
            wp = wp.next;
        }
        return head;
    }*/

    // second solution

    int size(ListNode head){
        ListNode wp = head;
        int res = 0;
        while(wp != null){
            res++;
            wp = wp.next;
        }
        return res;
    }

    public ListNode reverseList(ListNode head) {
        int listSize = size(head);

        // init array with all zeros
        int [] array = new int[size(head)];
        Arrays.fill(array, 0);

        // write elements in the reverse order to the array
        ListNode wp = head;
        int i = 0;
        while(wp!=null){
            array[listSize - i -1] = wp.val;
            i++;
            wp = wp.next;
        }

        // write elements to the list
        wp = head;
        i = 0;
        while(wp!=null){
            wp.val = array[i];
            i++;
            wp = wp.next;
        } 
        return head;
    }
}